### PR TITLE
sidecarset hot upgrade proposal

### PR DESCRIPTION
**Pull request by zmberg**
### Ⅰ. Describe what this PR does
sidecarset hot upgrade proposal

### CODE REVIEW

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+![sidecarset hotupgrade_injection](../../docs/img/sidecarset_hotupgrade_injection.png)
+
+### Hot Upgrade
+
+The SidecarSet Controller breaks down the hot upgrade process of the sidecar container into three steps:
+1. Upgrade: upgrade the empty container to the new version of the sidecar container, such as envoy-2.Image = envoy:1.17.0
+2. Migration: the process completes the state migration of stateful container, which needs to be done by the business itself according to its own container characteristics.
+PostStartHook completes the migration of the above process.
+(**Note: PostStartHook must be block during the migration, and exit when migration complete.**)
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:87]: PostStartHook must block during the migration

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+3. Reset: the step resets the old version sidecar container into empty container, such as envoy-1.Image = empty:1.0
+
+The above is the complete hot upgrade process. If a Pod needs to be hot upgraded several times, the above three steps can be repeated.
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:90]: please give more detail or suggestion about
1. how to determine the version of two upgrading sidecar containers
2. how two containers exchange states, through shared mounts, shared memory 
3. how to configure the resource request and limits of hot upgrade containers
4. what is the pod states we expect during the hot upgrade process, does it involves any updates on the service endpoints

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+$ kubectl describe pods test-pod
+  Name:         test-pod
+  Namespace:    default
+  Status:       Running
+  Containers:
+    nginx-sidecar-1:
+      Image:          empty:1.0.0
+      State:          Running
+        Started:      Tue, 09 Mar 2021 16:12:11 +0800
+    nginx-sidecar-2:
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:199]: the sample doesn't include SIDECARSET_VERSION variable

**Review by furykerry — CHANGES_REQUESTED** [review]

File: docs/proposals/20210305-sidecarset-hotupgrade.md (hunk 1 above)
Comment (Fei-Guo) [reply docs/proposals/20210305-sidecarset-hotupgrade.md:87]: In addition, you need to specify which container's PostStartHook you are referring to here.

File: docs/proposals/20210305-sidecarset-hotupgrade.md (hunk 2 above)
Comment (Fei-Guo) [reply docs/proposals/20210305-sidecarset-hotupgrade.md:90]: In addition, please also explain what will happen if any error happens during the migration process.

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+We can divide fd into two types: ListenFd and ConnectionFd, ListenFd is easier to migrate due to its stateless property, while ConnectionFd is not easy to complete the migration, the recommended migration solution is to migrate ListenFd, ConnectionFd "gracefully offline" and re-establish the connection to the new service process at the next request.
+
+**3. Exception Handling**
+
+Any abnormalities in the hot upgrade process will result in Pod ready=false, that is, the hot upgrade is lossy, so we only focus on whether the hot upgrade can end normally:
+- Old version of the container abnormal exit: kubelet will restart the exit container, PostStartHook script needs to be compatible with such scenarios, return to normal, the hot upgrade process will proceed normally
+- PostStartHook failure: kubelet will restart the new version of the container and execute the script PostStartHook failure until it succeeds
+
+#### Other Notes
+1. Because for the hot upgrade sidecar will be injected into two sidecar containers (resources.requests will be double), so recommended that resources.requests is not set or set to half
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:135]: For a single  hot upgrade sidecar in the sidecarset,  two sidecar containers  will be injected (resources.requests will double), so we recommend to leave the resources.requests unset, or set to half of the original  resources.requests

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+      value: '1097936677'
+  - name: envoy-2
+    env:
+    - name: SIDECARSET_VERSION
+      value: '1098297016'
+```
+Although the version of sidecar has been saved in the environment SIDECARSET_VERSION, it still needs to be negotiated between two sidecars to determine, and recommend to use shared Linux IPC, such as: TCP communication, shared memory, UDS, etc.
+The specific business can decide by itself according to its own technology stack.
+
+**NOTE: The next release will include environment SIDECARSET_VERSION_ALT( the other sidecar version ) in the container, so negotiation is no longer required**
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:121]: why not give the complete design in one shot , and give a implementation plan in the proposoal?

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+The above is the complete hot upgrade process. If a Pod needs to be hot upgraded several times, the above three steps can be repeated.
+
+![sidecarset hotupgrade](../../docs/img/sidecarset_hotupgrade.png)
+
+#### Migration Mechanism
+
+SidecarSet Hot upgrade provides an upgrade mechanism for stateful container. The specific state transfer process still needs to be implemented by the business itself.
+State migration is a very important step in the hot upgrade process, so i will describe some of the key points below:
+
+**1. Consultation Mechanism**
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:99]: Coordination Mechanism

**Review by furykerry — CHANGES_REQUESTED** [review]

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+The value of SIDECARSET_VERSION is the resourceVersion of the corresponding SidecarSet, so the larger one is the newer version compared with the two SIDECARSET_VERSION.
+As shown below, envoy-2 is the new version of sidecar:
+```
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: envoy-1
+    env:
+    - name: SIDECARSET_VERSION
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:112]: env in a pod is immutable once created, we should link the env to an annotations using downward api

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+We can divide fd into two types: ListenFd and ConnectionFd, ListenFd is easier to migrate due to its stateless property, while ConnectionFd is not easy to complete the migration, the recommended migration solution is to migrate ListenFd, ConnectionFd "gracefully offline" and re-establish the connection to the new service process at the next request.
+
+**3. Exception Handling**
+
+- **new version of sidecar container keeps failing to start (non-postStartHook execution failure)**: Image, or some other reasons could cause the repeating start-up failure of the new version of sidecar container. In this failure, the status of the pod will be not ready, but the old version of sidecar still remains running.
+under the circumstances, SidecarSet upgrades will be blocked here basing on the MaxUnavailable field unless a newer success version would be released.
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:176]: > SidecarSet upgrades will be blocked here basing on the MaxUnavailable field unless a newer success version would be released.
SidecarSet upgrades MAY be blocked due to the MaxUnavailable restriction,  unless a newer success version is released.

the blocking is not a must, maxUnavailable can tolerate some pod upgrade failure

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+- **failure during state migration (e.g. ListenFd migration succeeded, ConnectionFd graceful offline failed)**: In this case, the state resources (e.g. ListenFd) in the old version of the sidecar container will not be automatically restored, and the kubelet will restart the sidecar container at a later time.
+In this scenario, it is recommended that PostStartHook perform state migration no longer, but performs initialization operations to complete the start of the new version sidecar.
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:178]: > In this scenario, it is recommended that PostStartHook perform state migration no longer, but performs initialization operations to complete the start of the new version sidecar.
In this scenario,  PostStartHook MAY skip the state migration, and performs initialization operations to complete the start of the new version sidecar.

File: docs/proposals/20210305-sidecarset-hotupgrade.md [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+- **old version of sidecar container abnormal exit**: First of all, kubelet will restart the exit container, and after restarting, the sidecar container will firstly determine whether it’s in an old version. If yes, then it shouldn’t do any listening or other operations after starting except running.
```
Comment (furykerry) [inline docs/proposals/20210305-sidecarset-hotupgrade.md:179]: > old version of sidecar container abnormal exit
old version of sidecar container exit abnormally

> kubelet will restart the exit container
kubelet will restart the exited container 

> the sidecar container will firstly determine whether it’s in an old version. If yes, then it shouldn’t do any listening or other operations after starting except running.
the sidecar container will first determine whether it’s of old version. If so, it should just keep running but do nothing actually (such as listening for connections)

**Review by furykerry — CHANGES_REQUESTED** [review]

**Review by furykerry — APPROVED** [review]

**Pull request merged**